docs/2.10.0/docs/concepts/ledger-model/upgrade-model-in-depth.rst

```diff
@@ -0,0 +1,1680 @@
+      package *p2* is a valid upgrade of package *p1*, else *p2* and
+      *p1* are the exact same package.
+
+It is worth noting that even when *t2* upgrades *t1*, *r2* only upgrades
+*r1* provided that package *p2* is a valid upgrade of package *p1* as a
+whole.
+
+.. _examples-9:
+
+**Examples**
+
+In these examples we assume the existence of packages ``q-1.0.0`` and
+``q-2.0.0`` with LF version 1.16, and that the latter is a valid upgrade of
+the former.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - In ``q-1.0.0``:
+     - In ``q-2.0.0``:
+
+   * - .. code-block:: daml
+
+            module Dep where
+     
+            data U = C1
+            data V = V
+     
+     - :.. code-block:: daml
+     
+            module Dep where
+     
+            data U = C1 | C2
+            data V = V
+     
+Then below, the module on the right is a valid upgrade of the module on
+the left.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - .. code-block:: daml
+
+            module Main where
+     
+            -- imported from q-1.0.0
+            import qualified Dep
+     
+            data T = T Dep.U
+     
+     - .. code-block:: daml
+
+            module Main where
+     
+            -- imported from q-2.0.0
+            import qualified Dep
+     
+            data T = T Dep.U
+     
+However below, the module on the right is **not** a valid upgrade of the
+module on the left because ``Dep.V`` on the right belongs to package ``q-1.0.0``
+which is not a valid upgrade of package ``p-2.0.0``, even though the two
+definitions of ``V`` are the same.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - .. code-block:: daml
+
+            module Main where
+     
+            -- imported from q-2.0.0
+            import qualified Dep
+     
+            data T = T Dep.V
+
+     - .. code-block:: daml
+
+            module Main where
+     
+            -- imported from q-1.0.0
+            import qualified Dep
+     
+            data T = T Dep.V
+
+Suppose now that q-1.0.0 and q-2.0.0 are both compiled to LF version
+1.15 (which does not support upgrades). Then below, the module on the
+right is **not** a valid upgrade of the module on the left because the
+references to U on each side resolve to packages with different IDs.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - .. code-block:: daml
+
+            module Main where
+     
+            -- imported from q-1.0.0
+            import qualified Dep
+     
+            data T = T Dep.U
+     
+     - .. code-block:: daml
+
+            module Main where
+     
+            -- imported from q-2.0.0
+            import qualified Dep
+     
+            data T = T Dep.U
+
+Data Types - Parameterized Data Types
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Parameterized data types are considered serializable. That is, parameterized
```

Comment:
> Parameterized data types are considered serializable. 

This does not hold in this generality, does it? The assumption is that the parameters are considered serializable in the context of the serializability check.

Reply:
It looks like qualifying the sentence with "For the purpose of upgrade validation, parameterized data types are considered serializable, and thus checked for compatibility."

Reply:
I had this exact same sentence but Remy suggested I removed the "for the purpose of" part. This is because it actually holds generally: we do consider parameterized types serializable. But it's kind of a technicality because you only care about the serializability of these type constructors when fully applied in practice. 

@remyhaemmerle-da what do you think?

Reply:
As discussed, I just removed that paragraph. I have also since them added alpha equivalence so I reworked the examples.